p3387.tex

```diff
@@ -128,21 +132,23 @@ \subsection{Coroutine-ness is an implementation detail}
+%IDS I am not sure what conditions you envisage here;  A coroutine could suspend at the initial suspend and then be destroyed before the body was executed - but that would still return the coroutine object to the caller _before_ any such destruction.  Perhaps you envisage some fiendish implementation of an awaiter that could somehow do this? (at the moment I can't think of one - but ...)
 
 This lack of a directly visible connection between function declaration and function body may seem strange and unintuitive to users not familiar with coroutines; but it is a fundamental part of their design. Adding \tcode{pre} and \tcode{post} to the function declaration does not change this design and does not make things any more complex than they already are. This key insight allows us to move past ``postpone until we have a more complete picture'' and propose a coherent design for \tcode{pre} and \tcode{post} applied to coroutines.
 
 \subsection{Contract assertions on the suspend and resume points}
 
-In addition to contract assertions on the function interface, one might want to add contract assertions to the extended coroutine interface to specify preconditions and postconditions on its suspension and resumption, i.e. assertions on the values yielded from the coroutine and the state of the program at the suspend and resume points. However, since it is not visible on the function declaration whether a function is a coroutine, contract assertions on this extended interface cannot be specified on the function declaration, at least not without significantly modifying the design of coroutines, which we do not contemplate in this paper.
+??? In addition to contract assertions on the function interface, one might want to add contract assertions to the extended coroutine interface to specify preconditions and postconditions on its suspension and resumption, i.e. assertions on the values yielded from the coroutine and the state of the program at the suspend and resume points. However, since it is not visible on the function declaration whether a function is a coroutine, contract assertions on this extended interface cannot be specified on the function declaration, at least not without significantly modifying the design of coroutines, which we do not contemplate in this paper.
+%IDS I will send separate email about this - but I am begining to think that we (collectively) are over-thinking the internals - I can construct non-coroutine examples that have all these properties (with the exception of being a single-threaded impl).
```

Comment:
This paragraph is to address concerns of people who suggested that pre/post on a coroutine should somehow talk about the values that a coroutine yields/awaits on, and to tell them that this could be done but that would be a different language feature that we're not proposing. If you don't like this paragraph, do you have an idea how we could convey this better? Thanks

Reply:
I sent the separate mail - the essence is that the "wider interface" of a coroutine is, I suspect, mythical - one can construct regular functions (albeit requiring threading) with all the same properties and funkiness.  IMO this is a distraction - perhaps we can find a way to say that .. (i don't have a solution right now - but can think more on it).

Reply:
I don't think I received an email. What should we do with the paper, now? Remove the entire subsection? Or make certain edits to it? Which ones?

Reply:
"On the subject of a "wider" coroutine interface"
likewise, it will have to be tomorrow - I don't suppose we can remove the section if it addresses a voiced concern - but we could try to minimise the distraction.

Reply:
Yes that sounds like a good approach. I changed my mind - I'll leave the PR open until tomorrow morning and I'll wait for a resolution of both open issues before merging the PR and sharing the paper with SG21. Thanks again.

Reply:
Unfortunately to explain better why this is a distraction would involve more words - and therefore be a bigger distraction...

..... how about we just strengthen the comment a bit?:
"
We believe that most, if not all, initial use cases for such assertions can already be accomplished with the existing functionality in \cite{P2900R8} by using \tcode{contract_assert} inside the coroutine body or by applying \tcode{pre} and/or \tcode{post} to the various customisation functions in the promise and awaiter types. This latter approach is described in more detail in \cite{P3251R0}. A novel language feature to support this functionality is therefore  not required for a viable Contracts MVP
"